adapters/metax/metax.go

```diff
@@ -0,0 +1,196 @@
+	errs := make([]error, 0, len(request.Imp))
+
+	// split impressions
+	reqDatas := make([]*adapters.RequestData, 0, len(request.Imp))
+	for i := range request.Imp {
+		imp := &request.Imp[i]
+		impCopy := *imp
+		requestCopy := *request
+
+		metaxExt, err := parseBidderExt(imp)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if err := preprocessImp(&impCopy); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		endpoint, err := a.getEndpoint(metaxExt)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		requestCopy.Imp = []openrtb2.Imp{impCopy}
+		reqJSON, err := json.Marshal(requestCopy)
+		if err != nil {
+			errs = append(errs, err)
+			return nil, errs
+		}
+
+		headers := http.Header{}
+		headers.Add("Content-Type", "application/json;charset=utf-8")
+		headers.Add("Accept", "application/json")
+		reqDatas = append(reqDatas, &adapters.RequestData{
+			Method:  "POST",
+			Uri:     endpoint,
+			Body:    reqJSON,
+			Headers: headers,
+			ImpIDs:  openrtb_ext.GetImpIDs(requestCopy.Imp),
+		})
+	}
+
+	return reqDatas, errs
+}
+
+func (a *adapter) MakeBids(bidReq *openrtb2.BidRequest, reqData *adapters.RequestData, respData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if adapters.IsResponseStatusCodeNoContent(respData) {
+		return nil, nil
+	}
+
+	if err := adapters.CheckResponseStatusCodeForErrors(respData); err != nil {
+		return nil, []error{err}
+	}
+
+	var bidResp openrtb2.BidResponse
+	if err := json.Unmarshal(respData.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	// additional no content check
+	if len(bidResp.SeatBid) == 0 || len(bidResp.SeatBid[0].Bid) == 0 {
+		return nil, nil
+	}
+
+	resp := adapters.NewBidderResponseWithBidsCapacity(len(bidResp.SeatBid[0].Bid))
+	for _, sb := range bidResp.SeatBid {
+		for i := range sb.Bid {
+			bid := &sb.Bid[i]
+			bidType, err := getBidType(bid)
+			if err != nil {
+				return nil, []error{err}
+			}
+			resp.Bids = append(resp.Bids, &adapters.TypedBid{
+				Bid:     bid,
+				BidType: bidType,
+			})
+		}
+	}
+	return resp, nil
+}
+
+func (a *adapter) getEndpoint(ext *openrtb_ext.ExtImpMetaX) (string, error) {
+	params := macros.EndpointTemplateParams{
+		PublisherID: strconv.Itoa(ext.PublisherID),
+		AdUnit:      strconv.Itoa(ext.Adunit),
+	}
+	return macros.ResolveMacros(a.template, params)
+}
+
+func parseBidderExt(imp *openrtb2.Imp) (*openrtb_ext.ExtImpMetaX, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, err
+	}
+
+	var metaxExt openrtb_ext.ExtImpMetaX
+	if err := json.Unmarshal(bidderExt.Bidder, &metaxExt); err != nil {
+		return nil, errors.New("Wrong MetaX bidder ext")
+	}
+
+	return &metaxExt, nil
+}
+
+func preprocessImp(imp *openrtb2.Imp) error {
+	if imp == nil {
+		return errors.New("imp is nil")
+	}
+
+	if imp.Banner != nil {
+		imp.Banner = assignBannerSize(imp.Banner)
+	}
+
+	return nil
+}
```

Comment:
what it retur if we have "video, native or audio

Reply:
we handle the banner.format to avoid banner has formats but the width or height is zero. and only banner have this field so we only process banner here